#!/bin/bash

# assemble.sh generated by masurca
CONFIG_PATH="/work/raechill/clownfish_data/short_reads.config"
CMD_PATH="/work/raechill/MaSuRCA-3.2.4/bin/masurca"

# Test that we support <() redirection
(eval "cat <(echo test) >/dev/null" 2>/dev/null) || {
  echo >&2 "ERROR: The shell used is missing important features."
  echo >&2 "       Run the assembly script directly as './$0'"
  exit 1
}

# Parse command line switches
while getopts ":rc" o; do
  case "${o}" in
    c)
    echo "configuration file is '$CONFIG_PATH'"
    exit 0
    ;;
    r)
    echo "Rerunning configuration"
    exec perl "$CMD_PATH" "$CONFIG_PATH"
    echo "Failed to rerun configuration"
    exit 1
    ;;
    *)
    echo "Usage: $0 [-r] [-c]"
    exit 1
    ;;
  esac
done
set +e
# Set some paths and prime system to save environment variables
save () {
  (echo -n "$1=\""; eval "echo -n \"\$$1\""; echo '"') >> environment.sh
}
GC=
RC=
NC=
if tty -s < /dev/fd/1 2> /dev/null; then
  GC='\e[0;32m'
  RC='\e[0;31m'
  NC='\e[0m'
fi
log () {
  d=$(date)
  echo -e "${GC}[$d]${NC} $@"
}
fail () {
  d=$(date)
  echo -e "${RC}[$d]${NC} $@"
  exit 1
}
signaled () {
  fail Interrupted
}
trap signaled TERM QUIT INT
rm -f environment.sh; touch environment.sh

# To run tasks in parallel
#run_bg () {
#  semaphore -j $NUM_THREADS --id masurca_$$ -- "$@"
#}
#run_wait () {
#  semaphore -j $NUM_THREADS --id masurca_$$ --wait
#}
export PATH="/work/raechill/MaSuRCA-3.2.4/bin:/work/raechill/MaSuRCA-3.2.4/bin/../CA/Linux-amd64/bin:$PATH"
save PATH
export PERL5LIB=/work/raechill/MaSuRCA-3.2.4/bin/../lib/perl${PERL5LIB:+:$PERL5LIB}
save PERL5LIB
NUM_THREADS=16
save NUM_THREADS
log 'Processing pe library reads'
rm -rf meanAndStdevByPrefix.pe.txt
echo 'pe 155 18' >> meanAndStdevByPrefix.pe.txt
rename_filter_fastq 'pe' <(exec expand_fastq 'data/paired_end/SRR6255796_trimmed_1P' | awk '{if(length($0>250)) print substr($0,1,250); else print $0;}') <(exec expand_fastq 'data/paired_end/SRR6255798_trimmed_2P' | awk '{if(length($0>250)) print substr($0,1,250); else print $0;}' ) > 'pe.renamed.fastq'
echo 'se 150 7' >> meanAndStdevByPrefix.pe.txt
rename_filter_fastq 'se' <(exec expand_fastq 'data/SRR6266482.trimmed' | awk '{if(length($0>250)) print substr($0,1,250); else print $0;}') '' > 'se.renamed.fastq'

head -q -n 40000  pe.renamed.fastq se.renamed.fastq | grep --text -v '^+' | grep --text -v '^@' > pe_data.tmp
export PE_AVG_READ_LENGTH=`awk '{if(length($1)>31){n+=length($1);m++;}}END{print int(n/m)}' pe_data.tmp`
save PE_AVG_READ_LENGTH
log Average PE read length $PE_AVG_READ_LENGTH
KMER=`for f in pe.renamed.fastq se.renamed.fastq;do head -n 80000 $f |tail -n 40000;done | perl -e 'while($line=<STDIN>){$line=<STDIN>;chomp($line);push(@lines,$line);$line=<STDIN>;$line=<STDIN>}$min_len=100000;$base_count=0;foreach $l(@lines){$base_count+=length($l);push(@lengths,length($l));@f=split("",$l);foreach $base(@f){if(uc($base) eq "G" || uc($base) eq "C"){$gc_count++}}} @lengths =sort {$b <=> $a} @lengths; $min_len=$lengths[int($#lengths*.75)];  $gc_ratio=$gc_count/$base_count;$kmer=0;if($gc_ratio<0.5){$kmer=int($min_len*.7);}elsif($gc_ratio>=0.5 && $gc_ratio<0.6){$kmer=int($min_len*.5);}else{$kmer=int($min_len*.33);} $kmer++ if($kmer%2==0); $kmer=31 if($kmer<31); $kmer=151 if($kmer>151); print $kmer'`
save KMER
log Using kmer size of $KMER for the graph
KMER_J=$KMER
MIN_Q_CHAR=`cat pe.renamed.fastq se.renamed.fastq |head -n 50000 | awk 'BEGIN{flag=0}{if($0 ~ /^\+/){flag=1}else if(flag==1){print $0;flag=0}}'  | perl -ne 'BEGIN{$q0_char="@";}{chomp;@f=split "";foreach $v(@f){if(ord($v)<ord($q0_char)){$q0_char=$v;}}}END{$ans=ord($q0_char);if($ans<64){print "33\n"}else{print "64\n"}}'`
save MIN_Q_CHAR
echo MIN_Q_CHAR: $MIN_Q_CHAR
JF_SIZE=`ls -l *.fastq | awk '{n+=$5}END{s=int(n/50); if(s>200000000)printf "%.0f",s;else print "200000000";}'`
save JF_SIZE
perl -e '{if(int('$JF_SIZE')>200000000){print "WARNING: JF_SIZE set too low, increasing JF_SIZE to at least '$JF_SIZE', this automatic increase may be not enough!\n"}}'
log Creating mer database for Quorum
awk '{print substr($0,1,200)}' pe.renamed.fastq se.renamed.fastq | quorum_create_database -t 16 -s $JF_SIZE -b 7 -m 24 -q $((MIN_Q_CHAR + 5)) -o quorum_mer_db.jf.tmp /dev/stdin && mv quorum_mer_db.jf.tmp quorum_mer_db.jf
if [ 0 != 0 ]; then
  fail Increase JF_SIZE in config file, the recommendation is to set this to genome_size*coverage/2
fi

log Error correct PE.

quorum_error_correct_reads  -q $((MIN_Q_CHAR + 40)) --contaminant=/work/raechill/MaSuRCA-3.2.4/bin/../share/adapter.jf -m 1 -s 1 -g 1 -a 3 -t 16 -w 10 -e 3 -M  quorum_mer_db.jf pe.renamed.fastq se.renamed.fastq --no-discard -o pe.cor --verbose 1>quorum.err 2>&1 || {
  mv pe.cor.fa pe.cor.fa.failed && fail Error correction of PE reads failed. Check pe.cor.log.
}


if [ -s ESTIMATED_GENOME_SIZE.txt ];then
ESTIMATED_GENOME_SIZE=`head -n 1 ESTIMATED_GENOME_SIZE.txt`
else
log Estimating genome size.
jellyfish count -m 31 -t 16 -C -s $JF_SIZE -o k_u_hash_0 pe.cor.fa
export ESTIMATED_GENOME_SIZE=`jellyfish histo -t 16 -h 1 k_u_hash_0 | tail -n 1 |awk '{print $2}'`
echo $ESTIMATED_GENOME_SIZE > ESTIMATED_GENOME_SIZE.txt
fi
save ESTIMATED_GENOME_SIZE
echo "Estimated genome size: $ESTIMATED_GENOME_SIZE"

log Creating k-unitigs with k=$KMER
create_k_unitigs_large_k -c $(($KMER-1)) -t 16 -m $KMER -n $(($ESTIMATED_GENOME_SIZE*2)) -l $KMER -f `perl -e 'print 1/'$KMER'/1e5'` pe.cor.fa  | grep --text -v '^>' | perl -ane '{$seq=$F[0]; $F[0]=~tr/ACTGactg/TGACtgac/;$revseq=reverse($F[0]); $h{($seq ge $revseq)?$seq:$revseq}=1;}END{$n=0;foreach $k(keys %h){print ">",$n++," length:",length($k),"\n$k\n"}}' > guillaumeKUnitigsAtLeast32bases_all.fasta.tmp && mv guillaumeKUnitigsAtLeast32bases_all.fasta.tmp guillaumeKUnitigsAtLeast32bases_all.fasta
if [[ $KMER -eq $KMER_J ]];then
ln -s guillaumeKUnitigsAtLeast32bases_all.fasta guillaumeKUnitigsAtLeast32bases_all.jump.fasta
else
log Creating k-unitigs with k=$KMER_J
create_k_unitigs_large_k -c $(($KMER_J-1)) -t 16 -m $KMER_J -n $(($ESTIMATED_GENOME_SIZE*2)) -l $KMER_J -f `perl -e 'print 1/'$KMER_J'/1e5'` pe.cor.fa  | grep --text -v '^>' | perl -ane '{$seq=$F[0]; $F[0]=~tr/ACTGactg/TGACtgac/;$revseq=reverse($F[0]); $h{($seq ge $revseq)?$seq:$revseq}=1;}END{$n=0;foreach $k(keys %h){print ">",$n++," length:",length($k),"\n$k\n"}}' > guillaumeKUnitigsAtLeast32bases_all.jump.fasta.tmp && mv guillaumeKUnitigsAtLeast32bases_all.jump.fasta.tmp guillaumeKUnitigsAtLeast32bases_all.jump.fasta 
fi


log 'Computing super reads from PE '
rm -rf work1
CA_DIR="CA";
createSuperReadsForDirectory.perl -l $KMER -mean-and-stdev-by-prefix-file meanAndStdevByPrefix.pe.txt -kunitigsfile guillaumeKUnitigsAtLeast32bases_all.fasta -t 16 -mikedebug work1 pe.cor.fa 1> super1.err 2>&1
if [[ ! -e work1/superReads.success ]];then
fail Super reads failed, check super1.err and files in ./work1/
fi
create_sr_frg.pl < work1/superReadSequences.fasta 2>/dev/null | fasta2frg.pl super >  superReadSequences_shr.frg.tmp && mv superReadSequences_shr.frg.tmp superReadSequences_shr.frg

echo 1 > PLOIDY.txt
log 'Celera Assembler'
rm -rf CA
TOTAL_READS=`cat  *.frg | grep -c --text '^{FRG' `
save TOTAL_READS
ovlRefBlockSize=`perl -e '$s=int('$TOTAL_READS'/8); if($s>100000){print $s}else{print "100000"}'`
save ovlRefBlockSize
ovlHashBlockSize=`perl -e '$s=int('$TOTAL_READS'/80); if($s>10000){print $s}else{print "10000"}'`
save ovlHashBlockSize
ovlCorrBatchSize=$ovlHashBlockSize
save ovlCorrBatchSize
ovlMerThreshold=`jellyfish histo -t 16 k_u_hash_0 | awk '{thresh=75;if($1>1) {dist+=$2;if(dist>int("'$ESTIMATED_GENOME_SIZE'")*0.98&&flag==0){if($1>thresh) thresh=$1;flag=1}}}END{print thresh}'`
echo ovlMerThreshold=$ovlMerThreshold

runCA ovlMerThreshold=$ovlMerThreshold gkpFixInsertSizes=0 cgwErrorRate=0.15 jellyfishHashSize=$JF_SIZE ovlRefBlockSize=$ovlRefBlockSize ovlHashBlockSize=$ovlHashBlockSize ovlCorrBatchSize=$ovlCorrBatchSize stopAfter=consensusAfterUnitigger unitigger=bog -p genome -d CA merylThreads=16 frgCorrThreads=1 frgCorrConcurrency=16 cnsConcurrency=5 ovlCorrConcurrency=16 ovlConcurrency=16 ovlThreads=1 ovlMemory=4GB doFragmentCorrection=0 doOverlapBasedTrimming=0 doExtendClearRanges=0 ovlMerSize=30 superReadSequences_shr.frg      1> runCA1.out 2>&1

if [[ -e "CA/4-unitigger/unitigger.err" ]];then
  echo "Overlap/unitig success"
else
  fail Overlap/unitig failed, check output under CA/ and runCA1.out
fi

recompute_astat_superreads.sh genome CA $PE_AVG_READ_LENGTH work1/readPlacementsInSuperReads.final.read.superRead.offset.ori.txt

NUM_SUPER_READS=`cat superReadSequences_shr.frg  | grep -c --text '^{FRG' `
save NUM_SUPER_READS
if [ ! -e CA/overlapFilter.success ];then
( cd CA && tigStore -g genome.gkpStore -t genome.tigStore 2 -d layout -U | tr -d '-' | awk 'BEGIN{print ">unique unitigs"}{if($1 == "cns"){seq=$2}else if($1 == "data.unitig_coverage_stat" && $2>=5){print seq"N"}}' | jellyfish count -L 2 -C -m 30 -s $ESTIMATED_GENOME_SIZE -t 16 -o unitig_mers /dev/fd/0 &&  cat <(overlapStore -b 1 -e $NUM_SUPER_READS -d genome.ovlStore  | awk '{if($1<'$NUM_SUPER_READS' && $2<'$NUM_SUPER_READS') print $0}'|filter_overlap_file -t 16 <(gatekeeper -dumpfastaseq genome.gkpStore ) unitig_mers /dev/fd/0) <(overlapStore -d genome.ovlStore | awk '{if($1>='$NUM_SUPER_READS' || $2>='$NUM_SUPER_READS') print $1" "$2" "$3" "$4" "$5" "$6" "$7}')  |convertOverlap -b -ovl > overlaps.ovb &&  rm -rf 4-unitigger 5-consensus genome.tigStore && mkdir ovlStoreBackup && mv genome.ovlStore ovlStoreBackup && overlapStore -c genome.ovlStore -M 4096 -t 16 -g genome.gkpStore overlaps.ovb 1>overlapstore.err 2>&1 && rm overlaps.ovb && touch overlapFilter.success
)
fi

runCA ovlMerThreshold=$ovlMerThreshold gkpFixInsertSizes=0 cgwErrorRate=0.15 jellyfishHashSize=$JF_SIZE ovlRefBlockSize=$ovlRefBlockSize ovlHashBlockSize=$ovlHashBlockSize ovlCorrBatchSize=$ovlCorrBatchSize stopAfter=consensusAfterUnitigger unitigger=bog -p genome -d CA merylThreads=16 frgCorrThreads=1 frgCorrConcurrency=16 cnsConcurrency=5 ovlCorrConcurrency=16 ovlConcurrency=16 ovlThreads=1 ovlMemory=4GB doFragmentCorrection=0 doOverlapBasedTrimming=0 doExtendClearRanges=0 ovlMerSize=30 superReadSequences_shr.frg      1> runCA2.out 2>&1

if [[ -e "CA/5-consensus/consensus.success" ]];then
  echo "Unitig consensus success"
else
  echo "Fixing unitig consensus..."
  mkdir CA/fix_unitig_consensus
  ( cd CA/fix_unitig_consensus
    cp `which fix_unitigs.sh` .
    ./fix_unitigs.sh genome 
  )
fi

recompute_astat_superreads.sh genome CA $PE_AVG_READ_LENGTH work1/readPlacementsInSuperReads.final.read.superRead.offset.ori.txt
runCA cgwErrorRate=0.15 unitigger=bog -p genome -d CA cnsConcurrency=5 computeInsertSize=0 ovlMemory=4GB doFragmentCorrection=0 doOverlapBasedTrimming=0 doExtendClearRanges=0 ovlMerSize=30 1>runCA3.out 2>&1
if [[ -e "CA/9-terminator/genome.qc" ]];then
  echo "CA success"
else
  fail CA failed, check output under CA/ and runCA3.out
fi
TERMINATOR="9-terminator"
if [ -s $CA_DIR/9-terminator/genome.scf.fasta ];then
  NSCF=`grep --text '^>'  $CA_DIR/9-terminator/genome.scf.fasta |wc -l`
  NCTG=`grep --text '^>'  $CA_DIR/9-terminator/genome.ctg.fasta |wc -l`
  if [ $NCTG -eq $NSCF ];then
    log 'No gap closing possible.'
  else
    TERMINATOR="10-gapclose"
    if [ -s $CA_DIR/10-gapclose/genome.scf.fasta ];then
      log 'Gap closing done'
    else
      log 'Gap closing.'
      closeGapsLocally.perl --max-reads-in-memory 1000000000 -s 200000000 --Celera-terminator-directory $CA_DIR/9-terminator --reads-file 'pe.renamed.fastq' --reads-file 'se.renamed.fastq' --output-directory $CA_DIR/10-gapclose --min-kmer-len 17 --max-kmer-len $(($PE_AVG_READ_LENGTH-5)) --num-threads 16 --contig-length-for-joining $(($PE_AVG_READ_LENGTH-1)) --contig-length-for-fishing 200 --reduce-read-set-kmer-size 21 1>gapClose.err 2>&1
      if [[ -e "$CA_DIR/10-gapclose/genome.ctg.fasta" ]];then
        log 'Gap close success.'
      else
        fail Gap close failed, you can still use pre-gap close files under $CA_DIR/9-terminator/. Check gapClose.err for problems.
      fi
    fi
  fi
else
  fail "Assembly stopped or failed, see $CA_DIR.log"
fi
if [ -s $CA_DIR/$TERMINATOR/genome.scf.fasta ];then
  if [ ! -e $CA_DIR/filter_map.contigs.success ];then
    PLOIDY=`cat PLOIDY.txt`
    deduplicate_contigs.sh $CA_DIR genome 16 $PLOIDY $TERMINATOR && echo "Assembly complete, final scaffold sequences are in $CA_DIR/final.genome.scf.fasta"
  else
  log "Assembly complete, final scaffold sequences are in $CA_DIR/final.genome.scf.fasta"
  fi
else
  fail "Assembly stopped or failed, see $CA_DIR.log"
fi
log 'All done'
